### PR TITLE
expose api to let user control log output

### PR DESCRIPTION
**Pull request by evmar**
Add
  whisper_set_log_callback()
that lets user set a callback for log messages.

Change all the
  fprintf(stderr, ...)
to call via the above.

This is nice when embedding whisper.cpp into a larger program that doesn't want to print stuff to stderr.

### CODE REVIEW

**Comment by evmar** [comment]
Note: I also considered making a log macro that included the `__func__` bit automatically, lemme know if you think that's a good idea.  Also if it's literally named `log` then it shadows `std::log` the math function so it would need a different name.

File: whisper.cpp [hunk 1]
```diff
@@ -711,6 +711,20 @@ struct whisper_context {
...
 };
 
+static void whisper_default_log(const char * text) {
+    fprintf(stderr, "%s\n", text);
+}
+
+static whisper_log_callback whisper_log = whisper_default_log;
+
+static void log(const char * fmt, ...) {
+    char buf[256];
```
Comment (ggerganov) [inline whisper.cpp:721]:
```suggestion
    char buf[1024];
```

File: whisper.cpp [hunk 2]
```diff
@@ -711,6 +711,20 @@ struct whisper_context {
     std::string path_model; // populated by whisper_init_from_file()
 };
 
+static void whisper_default_log(const char * text) {
+    fprintf(stderr, "%s\n", text);
```
Comment (ggerganov) [inline whisper.cpp:715]: New lines must be explicitly logged

```suggestion
    fprintf(stderr, "%s", text);
```

File: whisper.cpp [hunk 3]
```diff
@@ -711,6 +711,20 @@ struct whisper_context {
...
+}
+
+static whisper_log_callback whisper_log = whisper_default_log;
+
+static void log(const char * fmt, ...) {
+    char buf[256];
+    va_list args;
+    va_start(args, fmt);
+    vsnprintf(buf, sizeof(buf), fmt, args);
+    whisper_log(buf);
```
Comment (ggerganov) [inline whisper.cpp:725]: Check if `whisper_log` is `nullptr`. If it is - do nothing

**Review by ggerganov — CHANGES_REQUESTED** [review]

**Comment by evmar** [comment]
OK, I applied your suggestions and rebased.

**Pull request merged**